### PR TITLE
Add API for taking a sequence of messages

### PR DESCRIPTION
**Pull request by mjcarroll**
This will allow users to take multiple messages in one take call if they
are available in the underlying middleware, versus having to make
multiple rmw_take calls in a row.

Signed-off-by: Michael Carroll <michael@openrobotics.org>

### CODE REVIEW

**Comment by mjcarroll** [comment]
Covering all branches:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/10084/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/5670/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/8203/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/9986/)

File: rmw/include/rmw/rmw.h [hunk 1]
```diff
@@ -710,6 +711,34 @@ rmw_take_with_info(
...
+ * In this case, only the currently available messages will be returned.
+ * The `taken` flag indicate the number of messages actually taken.
+ *
+ * \param[in] subscription The subscription object to take from.
+ * \param[in] count Number of messages to attempt to take.
+ * \param[out] message_sequence The sequence ROS message data on success.
+ * \param[out] message_info_sequence The seqeucne of additional message metadata on success.
+ * \param[out] taken Number of messages actually taken from subscription.
+ * \param[in] allocation Preallocated buffer to use (may be NULL).
+ * \return `RMW_RET_OK` if successful, or
```
Comment (iluetkeb) [inline rmw/include/rmw/rmw.h:728]: Please clarify whether RMW_RET_OK should be returned when fewer, or zero, messages were taken than requested.

File: rmw/include/rmw/rmw.h [hunk 2]
```diff
@@ -710,6 +711,34 @@ rmw_take_with_info(
...
+ * Take a sequence of ROS messgages from a given subscription.
+ *
+ * While `count` messages may be requested, fewer messages may be available on the subscription.
+ * In this case, only the currently available messages will be returned.
+ * The `taken` flag indicate the number of messages actually taken.
+ *
+ * \param[in] subscription The subscription object to take from.
+ * \param[in] count Number of messages to attempt to take.
+ * \param[out] message_sequence The sequence ROS message data on success.
+ * \param[out] message_info_sequence The seqeucne of additional message metadata on success.
```
[the remark this answers is not in the document]
Comment (wjwwood) [reply rmw/include/rmw/rmw.h:725]: Further more, should these sequences be initialized, zero initialized, or can they be reused from a previous take?

File: rmw/include/rmw/types.h [hunk 3]
```diff
@@ -337,13 +336,21 @@ typedef struct RMW_PUBLIC_TYPE rmw_message_info_t
   bool from_intra_process;
 } rmw_message_info_t;
 
+typedef struct RMW_PUBLIC_TYPE rmw_message_sequence_t
+{
+  void ** data;
+  size_t size;
+  size_t capacity;
+} rmw_message_sequence_t;
```
Comment (wjwwood) [inline rmw/include/rmw/types.h:344]: Out of curiosity, why move this from the header it was in? Ideally, imo, we should be moving things out of types.h into their own headers as we can.

File: rmw/src/message_sequence.c [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+    .size = 0u,
+    .capacity = 0u
+  };
+
+  return message_sequence;
+}
+
+rmw_ret_t
+rmw_message_sequence_init(
+  rmw_message_sequence_t * sequence, size_t size)
```
Comment (wjwwood) [inline rmw/src/message_sequence.c:31]: nitpick: why wrap this to a new line? And if you want to wrap it, why not wrap each argument to its own line?

File: rmw/src/message_sequence.c [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+{
+  RCUTILS_CHECK_ARGUMENT_FOR_NULL(sequence, RMW_RET_INVALID_ARGUMENT);
+
+  void * data = NULL;
+  if (size) {
+    data = malloc(sizeof(void *) * size);
```
Comment (wjwwood) [inline rmw/src/message_sequence.c:37]: This should use an allocator, so this function needs an allocator argument.

File: rmw/src/message_sequence.c [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+
+rmw_ret_t
+rmw_message_sequence_fini(
+  rmw_message_sequence_t * sequence)
+{
+  RCUTILS_CHECK_ARGUMENT_FOR_NULL(sequence, RMW_RET_INVALID_ARGUMENT);
+
+  if (sequence->data) {
+    assert(sequence->capacity > 0);
+    free(sequence->data);
```
Comment (wjwwood) [inline rmw/src/message_sequence.c:58]: This should use the allocator that was used to allocate it, so the allocator should probably be stored in the sequence.

File: rmw/src/message_sequence.c [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+
+rmw_ret_t
+rmw_message_info_sequence_init(
+  rmw_message_info_sequence_t * sequence, size_t size)
+{
+  RCUTILS_CHECK_ARGUMENT_FOR_NULL(sequence, RMW_RET_INVALID_ARGUMENT);
+
+  rmw_message_info_t * data = NULL;
+  if (size) {
+    data = malloc(sizeof(rmw_message_info_t) * size);
```
Comment (wjwwood) [inline rmw/src/message_sequence.c:90]: Same comments for these functions.

File: rmw/include/rmw/types.h (hunk 3 above)
Comment (mjcarroll) [reply rmw/include/rmw/types.h:344]: I'm fine with moving it back out into it's own header, I just didn't see any evidence of that happening currently.

File: rmw/src/message_sequence.c (hunk 4 above)
Comment (mjcarroll) [reply rmw/src/message_sequence.c:31]: I believe this was the change suggested by uncrustify.

Comment (wjwwood) [reply rmw/src/message_sequence.c:31]: Weird.

File: rmw/include/rmw/types.h (hunk 3 above)
Comment (wjwwood) [reply rmw/include/rmw/types.h:344]: We haven't been doing much moving out, but new types have been being added to their own headers, like this type and https://github.com/ros2/rmw/blob/master/rmw/include/rmw/topic_endpoint_info.h.

File: rmw/src/message_sequence.c [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+  rmw_message_sequence_t * sequence)
+{
+  RCUTILS_CHECK_ARGUMENT_FOR_NULL(sequence, RMW_RET_INVALID_ARGUMENT);
+
+  if (sequence->data) {
+    assert(sequence->capacity > 0);
+    free(sequence->data);
+    sequence->data = NULL;
+    sequence->size = 0;
+    sequence->capacity = 0;
```
Comment (jacobperron) [inline rmw/src/message_sequence.c:61]:
```suggestion
    sequence->size = 0u;
    sequence->capacity = 0u;
```

File: rmw/src/message_sequence.c [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+
+  rmw_message_info_t * data = NULL;
+  if (size) {
+    data = malloc(sizeof(rmw_message_info_t) * size);
+    if (!data) {
+      return RMW_RET_BAD_ALLOC;
+    }
+  }
+  sequence->data = data;
+  sequence->size = 0;
```
Comment (jacobperron) [inline rmw/src/message_sequence.c:96]:
```suggestion
  sequence->size = 0u;
```

File: rmw/src/message_sequence.c [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+  rmw_message_info_sequence_t * sequence)
+{
+  RCUTILS_CHECK_ARGUMENT_FOR_NULL(sequence, RMW_RET_INVALID_ARGUMENT);
+
+  if (sequence->data) {
+    assert(sequence->capacity > 0);
+    free(sequence->data);
+    sequence->data = NULL;
+    sequence->size = 0;
+    sequence->capacity = 0;
```
Comment (jacobperron) [inline rmw/src/message_sequence.c:112]:
```suggestion
    sequence->size = 0u;
    sequence->capacity = 0u;
```

File: rmw/test/test_message_sequence.cpp [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+
+#include "rcutils/allocator.h"
+
+#include "rmw/message_sequence.h"
+
+TEST(test_message_info_sequence, default_initialization) {
+  auto info_sequence = rmw_get_zero_initialized_message_info_sequence();
+
+  // Initialize the message sequence with a zero length
+  EXPECT_EQ(RMW_RET_OK, rmw_message_info_sequence_init(&info_sequence, 0));
```
Comment (jacobperron) [inline rmw/test/test_message_sequence.cpp:25]:
```suggestion
  EXPECT_EQ(RMW_RET_OK, rmw_message_info_sequence_init(&info_sequence, 0u));
```

File: rmw/test/test_message_sequence.cpp [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+  EXPECT_EQ(RMW_RET_OK, rmw_message_info_sequence_fini(&info_sequence));
+  EXPECT_EQ(0u, info_sequence.size);
+  EXPECT_EQ(0u, info_sequence.capacity);
+  EXPECT_EQ(nullptr, info_sequence.data);
+}
+
+TEST(test_message_info_sequence, initialization_with_size) {
+  auto info_sequence = rmw_get_zero_initialized_message_info_sequence();
+
+  EXPECT_EQ(RMW_RET_OK, rmw_message_info_sequence_init(&info_sequence, 5));
```
Comment (jacobperron) [inline rmw/test/test_message_sequence.cpp:39]:
```suggestion
  EXPECT_EQ(RMW_RET_OK, rmw_message_info_sequence_init(&info_sequence, 5u));
```

File: rmw/test/test_message_sequence.cpp [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+  EXPECT_EQ(0u, info_sequence.size);
+  EXPECT_EQ(0u, info_sequence.capacity);
+  EXPECT_EQ(nullptr, info_sequence.data);
+}
+
+TEST(test_message_sequence, default_initialization) {
+  auto message_sequence = rmw_get_zero_initialized_message_sequence();
+
+  // Initialize the message sequence with a zero length
+  EXPECT_EQ(RMW_RET_OK, rmw_message_sequence_init(&message_sequence, 0));
```
Comment (jacobperron) [inline rmw/test/test_message_sequence.cpp:53]:
```suggestion
  EXPECT_EQ(RMW_RET_OK, rmw_message_sequence_init(&message_sequence, 0u));
```

File: rmw/test/test_message_sequence.cpp [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+  EXPECT_EQ(RMW_RET_OK, rmw_message_sequence_fini(&message_sequence));
+  EXPECT_EQ(0u, message_sequence.size);
+  EXPECT_EQ(0u, message_sequence.capacity);
+  EXPECT_EQ(nullptr, message_sequence.data);
+}
+
+TEST(test_message_sequence, initialization_with_size) {
+  auto message_sequence = rmw_get_zero_initialized_message_sequence();
+
+  EXPECT_EQ(RMW_RET_OK, rmw_message_sequence_init(&message_sequence, 5));
```
Comment (jacobperron) [inline rmw/test/test_message_sequence.cpp:67]:
```suggestion
  EXPECT_EQ(RMW_RET_OK, rmw_message_sequence_init(&message_sequence, 5u));
```

File: rmw/test/test_message_sequence.cpp [hunk 15]
```diff
@@ -0,0 +1,75 @@
+// Copyright 2018 Open Source Robotics Foundation, Inc.
```
Comment (jacobperron) [inline rmw/test/test_message_sequence.cpp:1]:
```suggestion
// Copyright 2020 Open Source Robotics Foundation, Inc.
```

File: rmw/test/test_message_sequence.cpp [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+  EXPECT_EQ(nullptr, info_sequence.data);
+}
+
+TEST(test_message_info_sequence, initialization_with_size) {
+  auto info_sequence = rmw_get_zero_initialized_message_info_sequence();
+
+  EXPECT_EQ(RMW_RET_OK, rmw_message_info_sequence_init(&info_sequence, 5));
+  EXPECT_EQ(0u, info_sequence.size);
+  EXPECT_EQ(5u, info_sequence.capacity);
+
```
Comment (jacobperron) [inline rmw/test/test_message_sequence.cpp:42]:
```suggestion
EXPECT_NE(nullptr, info_sequence.data);
```

File: rmw/test/test_message_sequence.cpp [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+  EXPECT_EQ(nullptr, message_sequence.data);
+}
+
+TEST(test_message_sequence, initialization_with_size) {
+  auto message_sequence = rmw_get_zero_initialized_message_sequence();
+
+  EXPECT_EQ(RMW_RET_OK, rmw_message_sequence_init(&message_sequence, 5));
+  EXPECT_EQ(0u, message_sequence.size);
+  EXPECT_EQ(5u, message_sequence.capacity);
+
```
Comment (jacobperron) [inline rmw/test/test_message_sequence.cpp:70]:
```suggestion
EXPECT_NE(nullptr, info_sequence.data);
```

File: rmw/include/rmw/rmw.h (hunk 1 above)
Comment (mjcarroll) [reply rmw/include/rmw/rmw.h:735]: Clarified in a44b15a78991d01516e57c3f409ab5d7de4f473e

RMW_RET_OK will be returned in any case where there wasn't an error, even if fewer (or zero) messages were returned.  This is in line with the behavior of `rmw_take`, which will set the boolean flag accordingly without actually erroring.

File: rmw/include/rmw/rmw.h [hunk 18]
```diff
@@ -710,6 +711,34 @@ rmw_take_with_info(
...
+ * \param[out] taken Number of messages actually taken from subscription.
+ * \param[in] allocation Preallocated buffer to use (may be NULL).
+ * \return `RMW_RET_OK` if successful, or
+ * \return `RMW_RET_ERROR` if an unexpected error occurs.
```
[the remark this answers is not in the document]
Comment (mjcarroll) [reply rmw/include/rmw/rmw.h:739]: Clarified in a44b15a78991d01516e57c3f409ab5d7de4f473e

File: rmw/include/rmw/rmw.h [hunk 19]
```diff
@@ -710,6 +711,34 @@ rmw_take_with_info(
...
+/// Take multiple incoming messages from a subscription with additional metadata.
+/**
+ * Take a sequence of ROS messgages from a given subscription.
+ *
+ * While `count` messages may be requested, fewer messages may be available on the subscription.
+ * In this case, only the currently available messages will be returned.
+ * The `taken` flag indicate the number of messages actually taken.
+ *
+ * \param[in] subscription The subscription object to take from.
+ * \param[in] count Number of messages to attempt to take.
```
[the remark this answers is not in the document]
Comment (mjcarroll) [reply rmw/include/rmw/rmw.h:730]: Clarified in a44b15a78991d01516e57c3f409ab5d7de4f473e

File: rmw/include/rmw/rmw.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (mjcarroll) [reply rmw/include/rmw/rmw.h]: Clarified in a44b15a78991d01516e57c3f409ab5d7de4f473e

File: rmw/src/message_sequence.c (hunk 5 above)
Comment (mjcarroll) [reply rmw/src/message_sequence.c]: Added support for allocation here: 3d7d7a3

File: rmw/include/rmw/types.h (hunk 3 above)
Comment (mjcarroll) [reply rmw/include/rmw/types.h]: I have split the types out into their own file now.

File: rmw/src/message_sequence.c [hunk 20]
```diff
@@ -0,0 +1,128 @@
...
+#include "rmw/message_sequence.h"
+#include "rmw/types.h"
+
+rmw_message_sequence_t
+rmw_get_zero_initialized_message_sequence(void)
+{
+  static rmw_message_sequence_t message_sequence = {
+    .data = NULL,
+    .size = 0u,
+    .capacity = 0u
```
Comment (jacobperron) [inline rmw/src/message_sequence.c:24]:
```suggestion
    .capacity = 0u,
    .allocator = NULL
```

**Comment by mjcarroll** [comment]
CI incorporating all changes: 

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/10391/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/5904/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/8473/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/10287/)

**Comment by mjcarroll** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/10416/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/5922/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/8494/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/10305/)

**Comment by mjcarroll** [comment]
Linux and macOS are failing with Connext patching issues, similar to here: https://github.com/ros2/rmw_connext/pull/399#issuecomment-608509035

Windows went green too quickly.

**Comment by mjcarroll** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/10445/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/5939/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/8519/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/10326/)

**Comment by mjcarroll** [comment]
One more? 

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/10457/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/5947/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/8531/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/10338/)

**Comment by mjcarroll** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/10465/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/5951/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/8535/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/10348/)

**Comment by mjcarroll** [comment]
* macOS rebuild [Build Status](https://ci.ros2.org/job/ci_osx/8536/)

**Pull request merged**